cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -2572,6 +2572,249 @@
 (rule (lower (has_type (fits_in_64 ty) (uadd_overflow_trap a b tc)))
       (trap_if_overflow (add_with_flags_paired ty a b) tc))
 
+;;;; Helpers for `*_overflow` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; when true, sign extend, otherwise zero extend
+(decl put_in_reg_ext32 (Value bool) Reg)
+(rule (put_in_reg_ext32 val $true)
+      (put_in_reg_sext32 val))
+(rule (put_in_reg_ext32 val $false)
+      (put_in_reg_zext32 val))
+
+;; For values smaller than a register, we emit a normal op with both arguments
+;; zero/sign extended. We then check if the output is the same as itself zero/sign extended.
+(decl overflow_op_small (Type Value Value bool ALUOp) InstOutput)
+(rule (overflow_op_small ty a b sext alu_op)
+      (let ((extend ExtendOp (lower_extend_op ty sext))
+
+            ;; Instead of emitting two `{u,s}xt{b,h}` we do one as an instruction and
+            ;; the other as an extend operation in the alu_op.
+            ;;
+            ;; uxtb    a_ext, a
+            ;; alu_op  out, a_ext, b, {u,s}xtb
+            ;; cmp     out, out, {u,s}xtb
+            ;; cset    out_of, ne
+            (a_ext Reg (put_in_reg_ext32 a sext))
+            (out Reg (alu_rrr_extend alu_op ty a_ext b extend))
+            (out_of Reg (with_flags_reg 
+                  (cmp_extend (OperandSize.Size32) out out extend)
+                  (cset (Cond.Ne)))))
+      (output_pair
+            (value_reg out)
+            (value_reg out_of))))
+
+;; For register sized op's we just emit a op+cset, without further masking.
+;;
+;; op out, a, b
+;; cset out_of, cond
+(decl overflow_op_normal (Type Value Value ALUOp Cond) InstOutput)
+(rule (overflow_op_normal ty a b alu_op cond)
+      (let ((out ValueRegs
+              (with_flags
+                  (alu_rrr_with_flags_paired ty a b alu_op)
+                  (cset_paired cond))))
+      (output_pair
+            (value_regs_get out 0)
+            (value_regs_get out 1))))
+
+;; For 128bit integers we emit, for example, add+adcs+cset
+(decl overflow_op_128 (Value Value ALUOp ALUOp Cond) InstOutput)
+(rule (overflow_op_128 x y alu_op1 alu_op2 cond)
+      (let
+          ;; Get the high/low registers for `x`.
+          ((x_regs ValueRegs x)
+           (x_lo Reg (value_regs_get x_regs 0))
+           (x_hi Reg (value_regs_get x_regs 1))
+
+           ;; Get the high/low registers for `y`.
+           (y_regs ValueRegs y)
+           (y_lo Reg (value_regs_get y_regs 0))
+           (y_hi Reg (value_regs_get y_regs 1)))
+        ;; cannot use the with_flags helper here but it should be fine right now
```

Comment:
Let's extend the helper so we can -- it's the way we use to avoid subtle bugs from out-of-order `emit`s and unexpected side-effects.

Reply:
The problem is that currently `with_flags` returns a `ValueRegs` which can hold at max 2 values but the 128bit implementations for the overflow ops would need three return values.
So I could change the `with_flags` helper to return an `InstOutput` maybe and change all the places it's used to accomodate this but that would probably be quite a big change.
Or should I simply implement a new helper (e.g. `with_flags_extended`) that can return an `InstOutput`?
And should I use `InstOutput` or create a new type to handle more than one `ValueRegs`. Though I think the notion of `InstOuput` with multiple `ValueRegs` is already quite fitting.

Reply:
Hmm -- yeah, `with_flags_extended` seems like a better option here.